src/tools/wasm-shell.cpp

```diff
@@ -41,300 +40,380 @@ Name ASSERT_INVALID("assert_invalid");
+      actual = parseOperation(*s[1]);
+    } catch (const TrapException&) {
+      trapped = true;
+    } catch (const WasmException& e) {
+      std::cout << "[exception thrown: " << e << "]" << std::endl;
+      trapped = true;
+    }
+    assert(!trapped);
+    std::cerr << "seen " << actual << ", expected " << expected << '\n';
+    if (expected != actual) {
+      Fatal() << "unexpected, should be identical\n";
+    }
+  }
+
+  void parseModuleAssertion(Element& s) {
+    Module wasm;
+    wasm.features = FeatureSet::All;
+    std::unique_ptr<SExpressionWasmBuilder> builder;
+    auto id = s[0]->str();
+
+    bool invalid = false;
+    try {
+      SExpressionWasmBuilder(wasm, *s[1], IRProfile::Normal);
+    } catch (const ParseException&) {
+      invalid = true;
+    }
+
+    if (!invalid) {
+      // maybe parsed ok, but otherwise incorrect
+      invalid = !WasmValidator().validate(wasm);
+    }
+
+    if (!invalid && id == ASSERT_UNLINKABLE) {
+      // validate "instantiating" the mdoule
+      auto reportUnknownImport = [&](Importable* import) {
+        if (linkedInstances.count(import->module) == 0 ||
+            linkedInstances.at(import->module)
+                ->wasm.getExportOrNull(import->base) == nullptr) {
           std::cerr << "unknown import: " << import->module << '.'
                     << import->base << '\n';
           invalid = true;
-        };
-        ModuleUtils::iterImportedGlobals(wasm, reportUnknownImport);
-        ModuleUtils::iterImportedTables(wasm, reportUnknownImport);
-        ModuleUtils::iterImportedFunctions(wasm, [&](Importable* import) {
-          if (import->module == SPECTEST && import->base.startsWith(PRINT)) {
-            // We can handle it.
-          } else {
-            reportUnknownImport(import);
-          }
-        });
-        ElementUtils::iterAllElementFunctionNames(&wasm, [&](Name name) {
-          // spec tests consider it illegal to use spectest.print in a table
-          if (auto* import = wasm.getFunction(name)) {
-            if (import->imported() && import->module == SPECTEST &&
-                import->base.startsWith(PRINT)) {
-              std::cerr << "cannot put spectest.print in table\n";
-              invalid = true;
-            }
+        }
+      };
+      ModuleUtils::iterImportedGlobals(wasm, reportUnknownImport);
+      ModuleUtils::iterImportedTables(wasm, reportUnknownImport);
+      ModuleUtils::iterImportedFunctions(wasm, [&](Importable* import) {
+        if (import->module == SPECTEST && import->base.startsWith(PRINT)) {
+          // We can handle it.
+        } else {
+          reportUnknownImport(import);
+        }
+      });
+      ElementUtils::iterAllElementFunctionNames(&wasm, [&](Name name) {
+        // spec tests consider it illegal to use spectest.print in a table
+        if (auto* import = wasm.getFunction(name)) {
+          if (import->imported() && import->module == SPECTEST &&
+              import->base.startsWith(PRINT)) {
```

Comment:
Is `startswith` the correct check here or should the string be exactly `"print"`?

Reply:
This is a family of seven functions starting with `print`. You can take a look the spectest module definition in the reference interpreter's README: https://github.com/WebAssembly/spec/tree/master/interpreter#spectest-host-module